File: ObjC/ASTM30View-ObjC/ASTM30View-ObjC/ColorVectorParentViewController.m
//
//  ColorVectorParentViewController.m
//  ASTM30View-ObjC
//
//  Created by Inaba Mizuki on 2016/1/26.
//  Copyright © 2016年 Inaba Mizuki. All rights reserved.
//

#import "ColorVectorParentViewController.h"
#import "ASTM30CoordinateSpace.h"
#import "ASTM30PointsInfo.h"
#import "ASTM30ColorVectorViewController.h"
#import "Utilities.h"
#import "MZ.h"

@interface ColorVectorParentViewController ()
@property (readwrite, strong, nonatomic) ASTM30ColorVectorViewController* _tm30ViewController;
@end

@interface ColorVectorParentViewController (IB)
- (IBAction)triggerMask:(id)sender;
- (IBAction)takeSnapshot:(id)sender;
@end

@interface ColorVectorParentViewController (TestData)
- (NSArray<NSValue *> *)__referencePoints;
- (NSArray<NSValue *> *)__testSourcePoints;
@end



@implementation ColorVectorParentViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.view.backgroundColor = [UIColor grayColor];
    [self __testSetting];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    [super prepareForSegue:segue sender:sender];
    self._tm30ViewController = (ASTM30ColorVectorViewController *)segue.destinationViewController;
}

# pragma mark - Test

- (void)__testSetting {
    [self __addTestData];
}

- (void)__addTestData {
    self._tm30ViewController.testSourceName = @"TestSource";
    self._tm30ViewController.referenceName = @"Reference";

    mz_var(infoForReference, [[ASTM30PointsInfo alloc] initWithName:@"Reference"]);
    infoForReference.color = [UIColor blackColor];
    infoForReference.colorInMasked = [UIColor whiteColor];
    infoForReference.lineWidth = 4;
    infoForReference.points = ^{
        NSArray<NSValue*>* rawPoints = [self __referencePoints];
        NSMutableArray* points = [NSMutableArray array];

        for (int i = 0; i < rawPoints.count; i++) {
            NSString* key = [NSString stringWithFormat:@"A%d", i];
            [points addObject:[[ASTM30Point alloc] initWithKey:key
                                                         value:[rawPoints[i] CGPointValue]]];
        }

        return points;
    }();
    [self._tm30ViewController addPointsInfo:infoForReference];

    mz_var(infoForTestSource, [[ASTM30PointsInfo alloc] initWithName:@"TestSource"]);
    infoForTestSource.color = [UIColor redColor];
    infoForTestSource.colorInMasked = [UIColor clearColor];
    infoForTestSource.lineWidth = 4;
    infoForTestSource.points = ^{
        NSArray<NSValue*>* rawPoints = [self __testSourcePoints];
        NSMutableArray* points = [NSMutableArray array];

        for (int i = 0; i < rawPoints.count; i++) {
            NSString* key = [NSString stringWithFormat:@"A%d", i];
            [points addObject:[[ASTM30Point alloc] initWithKey:key
                                                         value:[rawPoints[i] CGPointValue]]];
        }

        return points;
    }();
    [self._tm30ViewController addPointsInfo:infoForTestSource];
}

@end

@implementation ColorVectorParentViewController (IB)

- (IBAction)triggerMask:(id)sender {
    ASTM30GraphicType type = (self._tm30ViewController.graphicType == ASTM30GraphicType_ColorVector)?
        ASTM30GraphicType_ColorDistortion :
        ASTM30GraphicType_ColorVector;

    [self._tm30ViewController setGraphicType:type];
}

- (IBAction)takeSnapshot:(id)sender {
    mz_var(snapshotImage, [Utilities imageWithView:self._tm30ViewController.view]);

    mz_var(imageData, UIImagePNGRepresentation(snapshotImage));

    mz_var(docPath, NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, true).firstObject);
    NSString* path = [NSString stringWithFormat:@"%@/%@.png", docPath, @"niya1"];

    BOOL success = [[NSFileManager defaultManager] createFileAtPath:path
                                                           contents:imageData
                                                         attributes:nil];
    MZLogMessageBool(@"Success", success);
    MZLog(@"Save to: %@", path);
}

@end

@implementation ColorVectorParentViewController (TestData)

- (NSArray<NSValue *> *)__referencePoints {
    return @[[NSValue valueWithCGPoint:CGPointMake(0.979482, 0.201533)],
             [NSValue valueWithCGPoint:CGPointMake(0.804753, 0.593609)],
             [NSValue valueWithCGPoint:CGPointMake(0.565791, 0.824548)],
             [NSValue valueWithCGPoint:CGPointMake(0.18508, 0.982723)],
             [NSValue valueWithCGPoint:CGPointMake(-0.0885433, 0.996072)],
             [NSValue valueWithCGPoint:CGPointMake(-0.516068, 0.856548)],
             [NSValue valueWithCGPoint:CGPointMake(-0.841959, 0.539542)],
             [NSValue valueWithCGPoint:CGPointMake(-0.993123, 0.117075)],
             [NSValue valueWithCGPoint:CGPointMake(-0.987714, -0.156275)],
             [NSValue valueWithCGPoint:CGPointMake(-0.849281, -0.527942)],
             [NSValue valueWithCGPoint:CGPointMake(-0.592092, -0.805871)],
             [NSValue valueWithCGPoint:CGPointMake(-0.184827, -0.982771)],
             [NSValue valueWithCGPoint:CGPointMake(0.112917, -0.993604)],
             [NSValue valueWithCGPoint:CGPointMake(0.597958, -0.801528)],
             [NSValue valueWithCGPoint:CGPointMake(0.827279, -0.561791)],
             [NSValue valueWithCGPoint:CGPointMake(0.983755, -0.179517)]];
}

- (NSArray<NSValue *> *)__testSourcePoints {
    return @[[NSValue valueWithCGPoint:CGPointMake(0.830428, 0.152285)],
             [NSValue valueWithCGPoint:CGPointMake(0.665421, 0.600832)],
             [NSValue valueWithCGPoint:CGPointMake(0.390513, 0.869416)],
             [NSValue valueWithCGPoint:CGPointMake(0.0299235, 1.03496)],
             [NSValue valueWithCGPoint:CGPointMake(-0.184857, 1.03375)],
             [NSValue valueWithCGPoint:CGPointMake(-0.521905,  0.903854)],
             [NSValue valueWithCGPoint:CGPointMake(-0.78017, 0.570425)],
             [NSValue valueWithCGPoint:CGPointMake(-0.87806, 0.12918)],
             [NSValue valueWithCGPoint:CGPointMake(-0.81089, -0.222912)],
             [NSValue valueWithCGPoint:CGPointMake(-0.631358, -0.654632)],
             [NSValue valueWithCGPoint:CGPointMake(-0.386599, -0.931485)],
             [NSValue valueWithCGPoint:CGPointMake(-0.0546757, -1.06064)],
             [NSValue valueWithCGPoint:CGPointMake(0.170136, -1.10126)],
             [NSValue valueWithCGPoint:CGPointMake(0.610058, -0.96514)],
             [NSValue valueWithCGPoint:CGPointMake(0.742329, -0.761882)],
             [NSValue valueWithCGPoint:CGPointMake(0.903637, -0.259266)]];
}

@end

File: Swift/ASTM30View-Swift/ASColorVectorViewProj/ColorVectorParentViewController.swift
//
//  ColorVectorParentViewController.swift
//  ASTM30View-Swift
//
//  Created by Inaba Mizuki on 2016/1/22.
//  Copyright © 2016年 Inaba Mizuki. All rights reserved.
//

import UIKit

class ColorVectorParentViewController: UIViewController {

    @IBOutlet weak var containerView: UIView!

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.blackColor()

        __testSetting()
    }

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        super.prepareForSegue(segue, sender: sender)
        _tm30ViewController = segue.destinationViewController as! ASTM30ColorVectorViewController
    }

    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransitionToSize(size, withTransitionCoordinator: coordinator)
        _tm30ViewController?.refresh()
    }


    // MARK: Private

    private var _tm30ViewController: ASTM30ColorVectorViewController! = nil
}


// MARK: IBActons

extension ColorVectorParentViewController {

    @IBAction func didTouchUpInsideMaskButton(button: UIButton) {
        let type =
            _tm30ViewController.graphicType == .ColorVector ?
                ASTM30GraphicType.ColorDistortion :
                ASTM30GraphicType.ColorVector
        _tm30ViewController.setGraphicType(type, animated: true)
    }

    @IBAction func didTouchUpInsideRemoveButton(button: UIButton) {
        _tm30ViewController.view.removeFromSuperview()
        _tm30ViewController.removeFromParentViewController()

        _tm30ViewController.view = nil
        _tm30ViewController = nil
    }
}


// Test
extension ColorVectorParentViewController {

    private func __testSetting() {
        __addTestData()
    }

    private func __addTestData() {
        _tm30ViewController.coordinateSpace = ASTM30CoordinateSpace(
            xMin: -1.2, xMax: 1.2, yMin: -1.2, yMax: 1.2
        )

        _tm30ViewController.testSourceName = "TestSource"
        _tm30ViewController.referenceName = "Reference"

        let infoForReference = ASTM30PointsInfo(name: "Reference")
        infoForReference.color = UIColor.blackColor()
        infoForReference.colorInMasked = UIColor.whiteColor()
        infoForReference.lineWidth = 4
        infoForReference.points = {
            let rawPoints = _referencePoints()
            var points = [ASTM30Point]()

            for var i = 0; i < rawPoints.count; i++ {
                let key = "A\(i)"
                points.append(ASTM30Point(key: key, value: rawPoints[i]))
            }

            return points
        }()
        _tm30ViewController.addPointsInfo(infoForReference)

        let infoForTestSource = ASTM30PointsInfo(name: "TestSource")
        infoForTestSource.color = UIColor.redColor()
        infoForTestSource.colorInMasked = UIColor.clearColor()
        infoForTestSource.lineWidth = 4
        infoForTestSource.points = {
            let rawPoints = _testSourcePoints()
            var points = [ASTM30Point]()

            for var i = 0; i < rawPoints.count; i++ {
                let key = "A\(i)"
                points.append(ASTM30Point(key: key, value: rawPoints[i]))
            }

            return points
        }()
        _tm30ViewController.addPointsInfo(infoForTestSource)
    }
}


// MARK: Test Datas
extension ColorVectorParentViewController {

    func _referencePoints() -> [CGPoint] {
        return [
            CGPoint(x: 0.979482, y: 0.201533),
            CGPoint(x: 0.804753, y: 0.593609),
            CGPoint(x: 0.565791, y: 0.824548),
            CGPoint(x: 0.18508, y: 0.982723),
            CGPoint(x: -0.0885433, y:  0.996072),
            CGPoint(x: -0.516068, y:  0.856548),
            CGPoint(x: -0.841959, y:  0.539542),
            CGPoint(x: -0.993123, y:  0.117075),
            CGPoint(x: -0.987714, y:  -0.156275),
            CGPoint(x: -0.849281, y:  -0.527942),
            CGPoint(x: -0.592092, y:  -0.805871),
            CGPoint(x: -0.184827, y:  -0.982771),
            CGPoint(x: 0.112917, y: -0.993604),
            CGPoint(x: 0.597958, y: -0.801528),
            CGPoint(x: 0.827279, y: -0.561791),
            CGPoint(x: 0.983755, y: -0.179517),
        ]
    }

    private func _testSourcePoints() -> [CGPoint] {
        return [
            CGPoint(x: 0.830428, y: 0.152285),
            CGPoint(x: 0.665421, y: 0.600832),
            CGPoint(x: 0.390513, y: 0.869416),
            CGPoint(x: 0.0299235, y: 1.03496),
            CGPoint(x: -0.184857, y:  1.03375),
            CGPoint(x: -0.521905, y:  0.903854),
            CGPoint(x: -0.78017, y:  0.570425),
            CGPoint(x: -0.87806, y:  0.12918),
            CGPoint(x: -0.81089, y:  -0.222912),
            CGPoint(x: -0.631358, y:  -0.654632),
            CGPoint(x: -0.386599, y:  -0.931485),
            CGPoint(x: -0.0546757, y:  -1.06064),
            CGPoint(x: 0.170136, y: -1.10126),
            CGPoint(x: 0.610058, y: -0.96514),
            CGPoint(x: 0.742329, y: -0.761882),
            CGPoint(x: 0.903637, y: -0.259266),
        ]
    }
}
